proposals/enum-entries.md

```diff
@@ -0,0 +1,174 @@
+
+Apart from that, most of the API being written leverages Collections API, not arrays, forcing users to manually convert arrays to lists.
+
+### Acknowledgements of the problem
+
+* It is considered a ["design bug"](http://mail.openjdk.java.net/pipermail/compiler-dev/2018-July/012242.html) in Java.
+* [JDK-8073381 need API to get enums values without creating a new arra...](https://bugs.openjdk.java.net/browse/JDK-8073381).
+* Scala 3 [attempted](https://github.com/lampepfl/dotty/issues/6620) to address this issue.
+* Graal [specifically optimizes](https://github.com/oracle/graal/issues/574) errorneous pattern `values()[idx]`.
+* [Memory-Hogging Enum.values() Method](https://dzone.com/articles/memory-hogging-enumvalues-method)
+
+### Examples of performance issues
+
+ * [HttpStatus.resolve allocates HttpStatus.values() once per invocation](https://github.com/spring-projects/spring-framework/issues/26842)
+ * [Kotlin standard library](https://github.com/JetBrains/kotlin/blob/92d200e093c693b3c06e53a39e0b0973b84c7ec5/libraries/stdlib/jvm/src/kotlin/text/CharCategoryJVM.kt#L170)
+ * [kotlinx.serializarion Enum deserializer]( https://github.com/Kotlin/kotlinx.serialization/issues/1372)
+ * [MySQL JDBC  Remove Enum.values() calls to avoid unnecessary array](https://github.com/Microsoft/mssql-jdbc/pull/1065)
+
+## Proposal
+
+The proposal that addresses all of the above is as follows:
+
+* Decommission of `Enum.values()` with the IDE assistance without deprecation.
+* Introduce property `entries: EnumEntries<E>` that returns an unmodifiable list of all enum entries.
+* For already compiled Kotlin enums and Java enums, a special mapping class that contains a pre-allocated list of entries is generated. 
+
+### Decommission of `Enum.values()`
+
+Enums have been existing since the very beginning of Kotlin, and since Java 1.5, released in 2004, meaning that the deprecation
+of `values` using Kotlin's standard [deprecation cycle](https://kotlinlang.org/docs/kotlin-evolution.html#incompatible-changes) 
+will create unnecessary disturbance in the ecosystem and will render a lot of already existing educational materials outdated.
+
+To avoid that, `values()` will be softly decommissioned via IDE assistance:
+
+* `values` will be de-prioritized and eventually removed from IDE auto-completion.
+* Soft warning with an intention to replace call to `values()` with call to `entries` will be introduced.
+* All the corresponding materials, such as Kotlin guides, J2K and tutorials will be adjusted to use `entries` API.
+* Eventually, we are going to re-visit this decision and decide on the further deprecation of `values` API.
+
+### `EnumEntries<E>` type
+
+Effectively, `entries` represents an immutable list of enum entries and can be represented as type `List<E>`.
+To have an ability to further extend Enum's API in a non-intrusive manner that does not involve code-generation
+for each enum, we expose a direct subtype of `List<E>` named `EnumEntries`.
```

Comment:
Makes sense to quote its full signature `sealed interface EnumEntries<E : Enum<E>> : List<E>` and explain why it's sealed.